apps/browser-proxy/src/index.ts

```diff
@@ -0,0 +1,195 @@
+import * as nodeNet from 'node:net'
+import * as https from 'node:https'
+import { PostgresConnection } from 'pg-gateway'
+import { WebSocketServer, type WebSocket } from 'ws'
+import makeDebug from 'debug'
+import * as tls from 'node:tls'
+import { extractDatabaseId, isValidServername } from './servername.ts'
+import { getTls } from './tls.ts'
+import { createStartupMessage } from './create-message.ts'
+import { extractIP } from './extract-ip.ts'
+
+const debug = makeDebug('browser-proxy')
+
+const tcpConnections = new Map<string, nodeNet.Socket>()
+const websocketConnections = new Map<string, WebSocket>()
+
+let tlsOptions = await getTls()
+
+// refresh the TLS certificate every week
+setInterval(
+  async () => {
+    tlsOptions = await getTls()
+    httpsServer.setSecureContext(tlsOptions)
+  },
+  1000 * 60 * 60 * 24 * 7
+)
+
+const httpsServer = https.createServer({
+  ...tlsOptions,
+  SNICallback: (servername, callback) => {
+    debug('SNICallback', servername)
+    if (isValidServername(servername)) {
+      debug('SNICallback', 'valid')
+      callback(null, tls.createSecureContext(tlsOptions))
+    } else {
+      debug('SNICallback', 'invalid')
+      callback(new Error('invalid SNI'))
+    }
+  },
+})
```

Comment:
This is for the web socket server right? I was surprised we care about SNI here. Are WS client also connecting through a unique URL?

Reply:
Ah okay I see we are 👍 

```typescript
const databaseHostname = `${databaseId}.${process.env.NEXT_PUBLIC_BROWSER_PROXY_DOMAIN}`
const ws = new WebSocket(`wss://${databaseHostname}`)
```